power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -21,114 +25,242 @@ namespace detail = power_grid_model::optimizer::detail;
+        add_to_edge(edges, edge_props, branch.to_node(), branch.from_node(), {unregulated_idx, 0});
+    }
+}
+
+template <typename... ComponentTypes, main_core::main_model_state_c State>
+inline auto add_edges(State const& state, RegulatedObjects const& regulated_objects, TrafoGraphEdges& edges,
+                      TrafoGraphEdgeProperties& edge_props) {
+    (add_edge<ComponentTypes>(state, regulated_objects, edges, edge_props), ...);
+}
+
+inline void add_to_edge(TrafoGraphEdges& edges, TrafoGraphEdgeProperties& edge_props, Idx const& start, Idx const& end,
+                        TrafoGraphEdge const& edge_prop) {
+    edges.emplace_back(start, end);
+    edge_props.emplace_back(edge_prop);
+}
+
 template <main_core::main_model_state_c State>
-inline auto build_transformer_graph(State const& /*state*/) -> TransformerGraph {
-    // TODO(nbharambe): implement
-    return {};
+inline auto retrieve_regulator_info(State const& state) -> RegulatedObjects {
+    RegulatedObjects regulated_objects;
+    for (auto const& regulator : state.components.template citer<TransformerTapRegulator>()) {
+        if (!regulator.status()) {
+            continue;
+        }
+        if (regulator.regulated_object_type() == ComponentType::branch) {
+            regulated_objects.transformers.emplace(regulator.regulated_object());
+        } else {
+            regulated_objects.transformers3w.emplace(regulator.regulated_object());
+        }
+    }
+    return regulated_objects;
+}
+
+template <main_core::main_model_state_c State>
+inline auto build_transformer_graph(State const& state) -> TransformerGraph {
+    TrafoGraphEdges edges;
+    TrafoGraphEdgeProperties edge_props;
+
+    const RegulatedObjects regulated_objects = retrieve_regulator_info(state);
+
+    add_edges<Transformer, ThreeWindingTransformer, Line, Link>(state, regulated_objects, edges, edge_props);
+
+    // build graph
+    TransformerGraph trafo_graph{boost::edges_are_unsorted_multi_pass, edges.cbegin(), edges.cend(),
+                                 edge_props.cbegin(),
+                                 static_cast<TrafoGraphIdx>(state.components.template size<Node>())};
+
+    BGL_FORALL_VERTICES(v, trafo_graph, TransformerGraph) { trafo_graph[v].is_source = false; }
+
+    if (state.components.template size<Source>() == 0) {
+        return trafo_graph;
+    }
+
+    // Mark sources
+    for (auto const& source : state.components.template citer<Source>()) {
+        // ignore disabled sources
+        trafo_graph[source.node()].is_source = source.status();
+    }
+
+    return trafo_graph;
 }
 
-inline auto process_edges_dijkstra(Idx v, std::vector<EdgeWeight>& edge_weight, std::vector<Idx2D>& edge_pos,
-                                   TransformerGraph const& graph) -> void {
+inline auto process_edges_dijkstra(Idx v, std::vector<EdgeWeight>& vertex_distances, TransformerGraph const& graph)
+    -> void {
```

Comment:
trailing return type `void` makes no sense. either keep it `auto` (no return type will be deduced) or just mark it `void` in the leading return type directly

```suggestion
inline void process_edges_dijkstra(Idx v, std::vector<EdgeWeight>& vertex_distances, TransformerGraph const& graph) {
```